mzLib/MassSpectrometry/MsDataFile.cs

```diff
@@ -146,29 +167,58 @@ public virtual IEnumerable<MsDataScan> GetMsScansInTimeRange(double firstRT, dou
             }
         }
 
+        /// <summary>
+        /// 
+        /// Performs a Binary Search and returns the closest Spectrum Number to the given Retention Time (RT).
+        /// If RT is negative or bigger than last RT recorded in file, an ArgumentException will be thrown.
+        /// 
+        /// </summary>
+        /// <param name="retentionTime"></param>
+        /// <returns></returns>
         public virtual int GetClosestOneBasedSpectrumNumber(double retentionTime)
         {
             if (!CheckIfScansLoaded())
+            {
                 LoadAllStaticData();
+            }
 
-            // TODO need to convert this to a binary search of some sort. Or if the data is indexedMZML see if the indices work better.
-            double bestDiff = double.MaxValue;
-            for (int i = 0; i < NumSpectra; i++)
+            if (ScansAndRetentionTime == null)
             {
-                double diff = Math.Abs(GetOneBasedScan(i + 1).RetentionTime - retentionTime);
-                if (diff > bestDiff)
-                    return i;
-                bestDiff = diff;
+                SetScansAndRetentionTimeDictionary();
```

Comment:
I feel like this is the only place where we should intialize the dictionary, because it's only used within this function